## Ordenación y clasificación
### 1. Ordenación de series por índice

In [1]:
import pandas as pd
import numpy as np

In [2]:
s = pd.Series([0,1,2,3,4], index = [3,1,5,0,4])
s

3    0
1    1
5    2
0    3
4    4
dtype: int64

In [3]:
#ordenamos por índice
s.sort_index()

0    3
1    1
3    0
4    4
5    2
dtype: int64

In [4]:
s.sort_index(ascending = False)

5    2
4    4
3    0
1    1
0    3
dtype: int64

**Si los index fueran letras, se seguiría el criterio Unicode**

### 2. Ordenación de series por valor

In [6]:
s = pd.Series([7,3,6,1,-4], index = ["a","b","c","d","e"])
s

a    7
b    3
c    6
d    1
e   -4
dtype: int64

In [7]:
##Ordenamos
s.sort_values() #tambien tiene el parámetro ascending

e   -4
d    1
b    3
c    6
a    7
dtype: int64

### 3. Ordenación de dataframes por índice

In [8]:
df = pd.DataFrame({"C":[-3,5,2],
                  "A":[1,0,3],
                  "D":[4,3,-4],
                  "B":[-2,3,1]},
                 index = ["c","a","b"])
df

C  A  D  B
c -3  1  4 -2
a  5  0  3  3
b  2  3 -4  1

In [9]:
df.sort_index() #Por filas

C  A  D  B
a  5  0  3  3
b  2  3 -4  1
c -3  1  4 -2

In [10]:
df.sort_index(axis = 1)#Por columnas

A  B  C  D
c  1 -2 -3  4
a  0  3  5  3
b  3  1  2 -4

In [11]:
#Para hacer una doble ordenación, tanto por filas como por columnas
df.sort_index().sort_index(axis = 1)

A  B  C  D
a  0  3  5  3
b  3  1  2 -4
c  1 -2 -3  4

In [12]:
#Podemos ordenar por valor
df.sort_values(by = ["A","C"])

C  A  D  B
a  5  0  3  3
c -3  1  4 -2
b  2  3 -4  1

In [13]:
df.sort_values(by = ["a","b"], axis = 1, ascending = False)

C  B  D  A
c -3 -2  4  1
a  5  3  3  0
b  2  1 -4  3

### 4. Clasificación de series
El método rank devuelve una serie conteniendo la clasificación o posición de cada valor de la serie original si fuesen ordenados de menor a mayor.

In [14]:
s = pd.Series([4,2,0,3,5], index = ["a","b","c","d","e"])
s

a    4
b    2
c    0
d    3
e    5
dtype: int64

In [15]:
s.rank()

a    4.0
b    2.0
c    1.0
d    3.0
e    5.0
dtype: float64

In [16]:
s = pd.Series([4,2,2,3,3,3,6], index =["a","b","c","d","e","f","g"])
s

a    4
b    2
c    2
d    3
e    3
f    3
g    6
dtype: int64

Si ordenamos los valores de menor a mayor:
2,2,3,3,3,4,5
Es decir, los valores 2 ocuparían las posiciones 1 y 2. Su valor medio es 1.5 que es el valor que asigna el método rank. 
Los valores 3 ocuparían las posiciones 3,4 y 5, cuyo valor medio es 4, y es lo que se muestra en rank. (3+4+5/3 = 4)

In [17]:
s.rank() #podemos porner method = "min" y en lugar de mostrar 1.5 daría 1.

a    6.0
b    1.5
c    1.5
d    4.0
e    4.0
f    4.0
g    7.0
dtype: float64

### 5. Clasificación de dataframes

In [18]:
sales = pd.DataFrame({"A":[3,3,1],
                     "B":[1,5,2],
                     "C":[3,7,2],
                     "D":[7,2,-1]},
                    index = ["Jan","Feb","Mar"])
sales

A  B  C  D
Jan  3  1  3  7
Feb  3  5  7  2
Mar  1  2  2 -1

In [20]:
sales.rank() #Para hacerlo por columnas axis = 1

A    B    C    D
Jan  2.5  1.0  2.0  3.0
Feb  2.5  3.0  3.0  2.0
Mar  1.0  2.0  1.0  1.0

La estructura devuelta es otro dataframe, y el eje por defecto en el que se calculan las clasificaciones es el eje 0 (eje vertical). 
* La primera columna está formada por las cifras 3, 3 y 1, y la clasificación es 2.5, 2.5 y 1 respectivamente, sabiendo que el 2.5 es la media de las posiciones 2 y 3 que dichas cifras ocuparían si la serie original se ordenase de menor a mayor.
* Se puede usar el método ascending como en las series.